docker/docker_image_src.go

```diff
@@ -30,15 +32,73 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	var imageSource *dockerImageSource
+	if registry == nil {
+		// If no registry was found during sysregistriesv2 lookup, we fallback
+		// to the standard behavior in pulling the fully qualified image from
+		// the remote.
```

Comment:
It would be a bit cleaner to create an artificial `registry` value here and fall through to the implementation below, to make sure the two won’t diverge; otherwise the mirror code path has a higher risk of breaking without anyone noticing.

Reply:
Adapted as suggested.